### PR TITLE
[ARROW] CommandResultExec should not trigger job

### PR DESCRIPTION
**Pull request by cfmcgrady**
### _Why are the changes needed?_

Before this PR:

截屏2023-04-13 下午2 34 55

After this PR:

截屏2023-04-13 下午2 33 19

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Review by pan3793 — APPROVED** [review]

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala [hunk 1]
```diff
@@ -177,6 +183,18 @@ object SparkDatasetHelper extends Logging {
...
+    assert(isRunnableCommand(command))
+    KyuubiArrowConverters.toBatchIterator(
+      // TODO: replace with `command.rows.iterator` once we drop Spark-3.1 support.
+      getRunnableCommandOutputRows(command).iterator,
+      command.schema,
+      SparkSession.active.sessionState.conf.arrowMaxRecordsPerBatch,
+      maxBatchSize,
+      -1,
+      SparkSession.active.sessionState.conf.sessionLocalTimeZone).toArray
+  }
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala:196]:
```suggestion
  lazy val commandResultExecRowsMethod = DynMethods.builder("rows")
      .impl("org.apache.spark.sql.execution.CommandResultExec")
      .build()

  def doCommandResultExec(command: SparkPlan): Array[Array[Byte]] = {
    assert(isRunnableCommand(command))
    KyuubiArrowConverters.toBatchIterator(
      // TODO: replace with `command.rows.iterator` once we drop Spark 3.1 support.
      commandResultExecRowsMethod.invoke[Seq[InternalRow]](command).iterator,
      command.schema,
      SparkSession.active.sessionState.conf.arrowMaxRecordsPerBatch,
      maxBatchSize,
      -1,
      SparkSession.active.sessionState.conf.sessionLocalTimeZone).toArray
  }
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala [hunk 2]
```diff
@@ -51,6 +54,9 @@ object SparkDatasetHelper extends Logging {
       doCollectLimit(collectLimit)
     case collectLimit: CollectLimitExec if collectLimit.limit < 0 =>
       executeArrowBatchCollect(collectLimit.child)
+    // TODO: replace with directly pattern match, once we drop Spark-3.1.x support.
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala:57]:
```suggestion
    // TODO: replace with pattern match once we drop Spark 3.1 support.
```

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/operation/SparkArrowbasedOperationSuite.scala [hunk 3]
```diff
@@ -278,6 +278,39 @@ class SparkArrowbasedOperationSuite extends WithSparkSQLEngine with SparkDataTyp
...
+          assert(resultSet.next())
+          assert(resultSet.getString("tableName") == "table_1")
+          KyuubiSparkContextHelper.waitListenerBus(spark)
+        }
+      }
+    }
+
+    if (SPARK_ENGINE_RUNTIME_VERSION < "3.2") {
+      // Note that before Spark-3.2, a LocalTableScan SparkPlan will be submitted, and the issue of
+      // preventing LocalTableScan from triggering a job submission was addressed in KYUUBI #4710.
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/operation/SparkArrowbasedOperationSuite.scala:304]:
```suggestion
      // Note that before Spark 3.2, a LocalTableScan SparkPlan will be submitted, and the issue of
      // preventing LocalTableScan from triggering a job submission was addressed in [KYUUBI #4710].
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala [hunk 4]
```diff
@@ -232,6 +248,33 @@ object SparkDatasetHelper extends Logging {
...
+    sparkPlan.getClass.getName == "org.apache.spark.sql.execution.CommandResultExec"
+  }
+
+  private lazy val commandResultExecRowsMethod = DynMethods.builder("rows")
+    .impl("org.apache.spark.sql.execution.CommandResultExec")
+    .build()
+
+  private def getRunnableCommandOutputRows(command: SparkPlan): Seq[InternalRow] = {
+    assert(isCommandResultExec(command))
+    commandResultExecRowsMethod.invoke[Seq[InternalRow]](command)
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala:275]: it's not complex, simply inline it.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala [hunk 5]
```diff
@@ -232,6 +248,33 @@ object SparkDatasetHelper extends Logging {
       .getOrElse(0)
   }
 
+  private def isCommandResultExec(sparkPlan: SparkPlan): Boolean = {
+    // scalastyle:off line.size.limit
+    // the CommandResultExec was introduced in SPARK-35378 (Spark-3.2.x), after SPARK-35378 the
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala:253]:
```suggestion
    // the CommandResultExec was introduced in SPARK-35378 (Spark 3.2), after SPARK-35378 the
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala [hunk 6]
```diff
@@ -177,6 +182,17 @@ object SparkDatasetHelper extends Logging {
     result.toArray
   }
 
+  def doCommandResultExec(command: SparkPlan): Array[Array[Byte]] = {
+    KyuubiArrowConverters.toBatchIterator(
+      // TODO: replace with `command.rows.iterator` once we drop Spark-3.1 support.
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala:187]:
```suggestion
      // TODO: replace with `command.rows.iterator` once we drop Spark 3.1 support.
```

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala (hunk 4 above)
Comment (cfmcgrady) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/sql/kyuubi/SparkDatasetHelper.scala]: updated.

**Review by cxzl25 — APPROVED** [review]

**Review by pan3793 — APPROVED** [review]

**Review by ulysses-you — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master/1.7

**Pull request closed without merging**